cng/pbkdf2.go

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//go:build windows
+// +build windows
+
+package cng
+
+import (
+	"errors"
+	"hash"
+	"unsafe"
+
+	"github.com/microsoft/go-crypto-winnative/internal/bcrypt"
+)
+
+func loadPBKDF2() (bcrypt.ALG_HANDLE, error) {
+	h, err := loadOrStoreAlg(bcrypt.PBKDF2_ALGORITHM, 0, "", func(h bcrypt.ALG_HANDLE) (interface{}, error) {
+		return h, nil
+	})
+	if err != nil {
+		return 0, err
+	}
+	return h.(bcrypt.ALG_HANDLE), nil
+}
+
+func PBKDF2(password, salt []byte, iter, keyLen int, h func() hash.Hash) ([]byte, error) {
+	ch := h()
+	hashID := hashToID(ch)
+	if hashID == "" {
+		return nil, errors.New("cng: unsupported hash function")
+	}
+	alg, err := loadPBKDF2()
+	if err != nil {
+		return nil, err
+	}
+	if len(password) == 0 {
+		// CNG doesn't support zero-length passwords.
+		password = make([]byte, 64) // 64 bytes is SHA512, the largest digest handled.
```

Comment:
This is a bit mysterious to me, probably because I don't have an intuition about what's happening internally. Why is a bunch of zeros equivalent to "no password" (or is it)? Why does it need to match the largest digest that the function's able to handle (or is this incidental)?

Reply:
> This is a bit mysterious to me, probably because I don't have an intuition about what's happening internally. Why is a bunch of zeros equivalent to "no password" (or is it)?

[RFC 2104, Section 2](https://www.rfc-editor.org/rfc/rfc2104#section-2) (the HMAC spec) I says that if the key is shorter than the block size, then it will be extended by appending zeros:

>  (1) append zeros to the end of K to create a B byte string
        (e.g., if K is of length 20 bytes and B=64, then K will be
         appended with 44 zero bytes 0x00)

Demonstration: https://play.golang.com/p/CgzDK_Z7dTV

> Why does it need to match the largest digest that the function's able to handle (or is this incidental)?

If we don't do it, CNG will do it for us. What CNG doesn't want is a NULL password, it fails if that happens, and `bcrypt.GenerateSymmetricKey` passes down a NULL password if it is empty.

I've refactored a little this part of the code in https://github.com/microsoft/go-crypto-winnative/pull/40/commits/81b780da8954c943e8d3eb9e6a573780614b712f. It now justs passes a 1-byte slice, hopefully is more understandable this way.

Reply:
Ah, thanks! Totally makes sense. I think a pointer to the spec would help, whether it passes one 0 or many 0s--it wasn't obvious to me what spec to look at that's governing this and without context, "0000" is definitely a different password than "0", or "". 😄 (Maybe it's ok to keep the comments light and targeted for those who already know the relevant specs, I'm not sure.)

Reply:
With https://github.com/microsoft/go-crypto-winnative/commit/81b780da8954c943e8d3eb9e6a573780614b712f changes I don't think it's necessary to document that because we are passing `0` as key length when password length is 0, even if we stack allocate one byte.

Reply:
I see, when I first saw that I assumed it was a mistake (hence other comment thread). I agree no explanation is needed if we're just passing a 0-length password.